test/blackbox/api/dds-pim/PubSubParticipant.hpp

```diff
@@ -144,6 +146,60 @@ class PubSubParticipant
         PubSubParticipant* participant_;
     };
 
+    class ParticipantListener : public eprosima::fastdds::dds::DomainParticipantListener
+    {
+        friend class PubSubParticipant;
+
+    public:
+
+        ParticipantListener(
+                PubSubParticipant* participant)
+            : participant_(participant)
+        {
+        }
+
+        ~ParticipantListener() = default;
+
+        void on_participant_discovery(
+                eprosima::fastdds::dds::DomainParticipant*,
+                eprosima::fastrtps::rtps::ParticipantDiscoveryInfo&& info)
+        {
+            if (participant_->on_discovery_ != nullptr &&
+                    info.status == eprosima::fastrtps::rtps::ParticipantDiscoveryInfo::DISCOVERED_PARTICIPANT)
+            {
+                std::unique_lock<std::mutex> lock(participant_->mutex_discovery_);
+                participant_->discovery_result_ |= participant_->on_discovery_(info);
+                participant_->cv_discovery_.notify_one();
+            }
```

Comment:
I'd remove `participant_matched()` and `participant_unmatched()`, and move there logic to the corresponding if branch. Furthermore, [documentation](https://en.cppreference.com/w/cpp/thread/condition_variable/notify_one) suggestions unlocking before notifying to avoid an unnecessary lock on the other side. Moreover, you can do `else if`.
```suggestion
            if (info.status == eprosima::fastrtps::rtps::ParticipantDiscoveryInfo::DISCOVERED_PARTICIPANT)
            {
                std::unique_lock<std::mutex> lock(participant_->mutex_discovery_);
                ++matched_;
                if (nullptr != participant_->on_discovery_)
                {
                    participant_->discovery_result_ |= participant_->on_discovery_(info);
                }
                lock.unlock();
                participant_->cv_discovery_.notify_one();
            }
```